api/service/projects_service.go

```diff
@@ -76,6 +81,24 @@ func (service *projectsService) CreateProject(ctx context.Context, project *mode
 		}
 	}
 
+	if service.webhookManager != nil {
+		err = service.webhookManager.InvokeWebhooks(ctx, ProjectCreatedEvent, project, func(p []byte) error {
+			// Expects webhook output to be a project object
+			var tmpproject models.Project
+			if err := json.Unmarshal(p, &tmpproject); err != nil {
+				return err
+			}
+			project, err = service.save(&tmpproject)
+			if err != nil {
+				return err
+			}
```

Comment:
Not really a question specific to this PR or the webhook implementation, but I was wondering how we should be handling scenarios whereby the webhooks responses are failures - should we even update the project in the first place? i.e. do not update the project at all if the webhooks are not called successfully vs do whatever the original method is supposed to do, save the project, calls the webhooks and if they fail, leave things as they are. 

Given our use own specific use-case(s) for these webhooks when updating projects, I think it's fine how we're handling such a scenario for now, but I'm wondering if there are scenarios whereby we might want to not do anything if the calls to the webhooks fail. What do you all think?

Reply:
The `onsuccess` callback method is only invoked if all the sync webhooks are successful. If any of the webhooks fail, the onsuccess callback is not invoked. If there is an error handler provided, the error handler will be invoked. Unsuccessful webhook calls are calls whose responses are invalid json objects, or http status not ok. Does that make things clearer?

Reply:
Oh I think that part which you explained is clear, what I'm trying to describe is if there are scenarios where we should be using these webhooks sometimes in a more atomic manner. To use this `CreateProject` function as an example, right now the following steps (roughly) occur when creating a project (any of these steps might fail and break midway):
1. Check if a reserved project name is used
2. Save the project
3. Update authorization policy if enabled
4. Call webhooks and save project again

I was wondering if we should be saving the project in 2 separate steps or if we should only save it once after the webhooks have been called successfully. If the webhooks serve to only add additional non-critical information to the project, then it's probably okay to save the project before the webhooks get triggered even if the webhook calls fail. The opposite would be to only save the project once all the webhooks successfully return a response (do not save anything before the webhooks return successfully).

On a totally unrelated note, should we also add this webhook calls to the `UpdateProject` service method? I believe that a user updating his/her project would probably also require the webhooks to be called, to do whatever needs to be done similarly to when a project is created for the first time.

Reply:
> I was wondering if we should be saving the project in 2 separate steps or if we should only save it once after the webhooks have been called successfully. If the webhooks serve to only add additional non-critical information to the project, then it's probably okay to save the project before the webhooks get triggered even if the webhook calls fail. The opposite would be to only save the project once all the webhooks successfully return a response (do not save anything before the webhooks return successfully).

I see what you mean. It depends on the use case, we could also have an event for PreProjectCreatedEvent and PostProjectCreatedEvent to differentiate between the 2. Whether post or pre depends on if the event itself has to take place first. For this, it could be moved to be called before the project is created, and in the callback we save the project to the db. The downside is that if any of the webhooks fail, project creation fails. I can make the change if that's preferred.

Reply:
Ahh I see no it's good as it is :) Thanks for the clarification on the potential options on how we can use the webhooks!